### ISSUE TITLE
TierPriceManagement: add() fails randomly because of invalid argument supplied for foreach()

### ISSUE THREAD

**Issue by HenriNijborg** [magento/magento2#30908]
File: app/code/Magento/Catalog/Model/Product/TierPriceManagement.php
Function: add()

Error `Invalid argument supplied for foreach()` on line ` foreach ($tierPrices as &$item)`

```php
        ...
        $tierPrices = $product->getData('tier_price');
        ...

        foreach ($tierPrices as &$item) {
            if ('all' == $customerGroupId) {
                $isGroupValid = ($item['all_groups'] == 1);
            } else {
                $isGroupValid = ($item['cust_group'] == $customerGroupId);
            }

            if ($isGroupValid && $item['website_id'] == $websiteIdentifier && $item['price_qty'] == $qty) {
                $item['price'] = $price;
                $found = true;
                break;
            }
        }
       ...
```

The behaviour is random. Strongly related to issue #24410 and bug fix https://github.com/magento/magento2/commit/114ac4b39a5818df2f1c91c07fd9b27cc97de7b1 

### Preconditions (*)
1. Magento 2.3.5-p1

### Steps to reproduce (*)
1. Add a tier price programmaticaly using the TierPriceManagement class.
2. `foreach ($tierPrices as &$item)` randomly throws `Invalid argument supplied for foreach()`

### Expected result (*)
1. $tierPrices should be checked for being null just as within bug fix https://github.com/magento/magento2/commit/114ac4b39a5818df2f1c91c07fd9b27cc97de7b1
2. Tier Price is added.

### Actual result (*)
1. `$tierPrices` isn't checked for being `null` causing the error. 
2. Tier Prices aren't added

### Suspicions
I think the result of a cached product has `tier_price: null` and the non cached product is `tier_price: []` as the behaviour is randomly and caching might be not for all products and not always valid anymore.

### Suggested fix
```php
public function add($sku, $customerGroupId, $price, $qty)
    {
        if (!is_float($price) && !is_int($price) && !\Zend_Validate::is((string)$price, 'Float')
            || !is_float($qty) && !is_int($qty) && !\Zend_Validate::is((string)$qty, 'Float')
            || $price <= 0
            || $qty <= 0
        ) {
            throw new InputException(__('The data was invalid. Verify the data and try again.'));
        }
        $product = $this->productRepository->get($sku, ['edit_mode' => true]);
        $websiteIdentifier = 0;
        $value = $this->config->getValue('catalog/price/scope', \Magento\Store\Model\ScopeInterface::SCOPE_WEBSITE);
        if ($value != 0) {
            $websiteIdentifier = $this->storeManager->getWebsite()->getId();
        }
        $found = false;

        $tierPrices = $product->getData('tier_price');
        if($tierPrices !== null) {
            foreach ($tierPrices as &$item) {
                if ('all' == $customerGroupId) {
                    $isGroupValid = ($item['all_groups'] == 1);
                } else {
                    $isGroupValid = ($item['cust_group'] == $customerGroupId);
                }

                if ($isGroupValid && $item['website_id'] == $websiteIdentifier && $item['price_qty'] == $qty) {
                    $item['price'] = $price;
                    $found = true;
                    break;
                }
            }
        }
        
        if (!$found) {
            $mappedCustomerGroupId = 'all' == $customerGroupId
                ? $this->groupManagement->getAllCustomersGroup()->getId()
                : $this->groupRepository->getById($customerGroupId)->getId();

            $tierPrices[] = [
                'cust_group' => $mappedCustomerGroupId,
                'price' => $price,
                'website_price' => $price,
                'website_id' => $websiteIdentifier,
                'price_qty' => $qty,
            ];
        }

        $product->setData('tier_price', $tierPrices);
        $errors = $product->validate();
        if (is_array($errors) && count($errors)) {
            $errorAttributeCodes = implode(', ', array_keys($errors));
            throw new InputException(
                __('Values in the %1 attributes are invalid. Verify the values and try again.', $errorAttributeCodes)
            );
        }
        try {
            $this->productRepository->save($product);
        } catch (\Exception $e) {
            if ($e instanceof TemporaryStateExceptionInterface) {
                // temporary state exception must be already localized
                throw $e;
            }
            throw new CouldNotSaveException(__("The group price couldn't be saved."));
        }
        return true;
    }
```

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [x] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by chiranjeevi-cj** [comment]
@magento I am working on this

**Comment by chiranjeevi-cj** [comment]
@engcom-Alfa Should I proceed with the suggested solution by @HenriNijborg ?

**Comment by HenriNijborg** [comment]
I think there is a second line that needs fixing:
`$product = $this->productRepository->get($sku, ['edit_mode' => true]);`

should be retrieving the products without using the cache as it will be used to compare the new rules to the existing ones. If the cache isn't valid then my suggested fix will cause no error and it will proceed and overide possibly tier prices which aren't in the cache but are in the db.

`$product = $this->productRepository->get($sku, ['edit_mode' => true, 'force_reload' => true]);` 
or 
`$product = $this->productRepository->get($sku, ['edit_mode' => true], null, true);`

I'm not a Magento 2 developer so I'm not sure which one it should be.

**Comment by HenriNijborg** [comment]
Yes, this issue is still relevant. Pull Request author @chiranjeevi-cj asked for help from @magento-engcom-team in related pull request https://github.com/magento/magento2/pull/30941

**Comment by vehgroshop** [comment]
What is the status of this issue? We would also like to have it fixed.

**Comment by PatrickJager** [comment]
@magento-engcom-team @chiranjeevi-cj 
What is the status on this issue? Did you manage to fix it?
We currently have another customer having this exact same issue. 
Some tierprices work some do not.

**Comment by engcom-Bravo** [comment]
Hi @HenriNijborg,

Thank you for reporting and collaboration.

Verified the issue on Magento 2.4-develop instance and the issue is not reproducible.Kindly refer the screenshots.


1.Added a tier price programmaticaly using the TierPriceManagement class.
2.foreach ($tierPrices as &$item) randomly throws Invalid argument supplied for foreach()


<img width="1359" alt="Screenshot 2023-01-27 at 5 54 13 PM" src="https://user-images.githubusercontent.com/51680745/215086849-538072b3-90d8-47d9-918a-24796f995e24.png">



The issue is no more reproducible.it is already fixed in 2.4-develop.

Thanks.



The isssue is no more [reproducible.It](http://reproducible.it/) is already fixed in 2.4-develop.

**Comment by engcom-Bravo** [comment]
Hi @HenriNijborg,


We have noticed that this issue has not been updated since long time. Hence we assume that this issue is fixed now, so we are closing it. Please feel to raise a fresh ticket or reopen this ticket if you need more assistance on this.

Thanks.
